HighlightUtilities/functions.php

```diff
@@ -196,56 +196,59 @@ function getLanguageDefinitionPath($name)
  *
  * @api
  *
+ * @since 9.18.1.6 `RuntimeException` and `UnexpectedValueException` can no longer be thrown.
  * @since 9.15.6.1
  *
  * @param string $html An HTML string generated by `Highlighter::highlight()`
  *
- * @throws \RuntimeException         when the DOM extension is not available
- * @throws \UnexpectedValueException when the given HTML could not be parsed
- *
  * @return string[]|false An array of lines of code as strings. False if an error occurred in splitting up by lines
  */
 function splitCodeIntoArray($html)
 {
-    if (!extension_loaded("dom")) {
-        throw new \RuntimeException("The DOM extension is not loaded but is required.");
-    }
-
     if (trim($html) === "") {
         return array();
     }
 
-    $dom = new \DOMDocument();
+    $queuedPrefix = '';
+    $regexWorkspace = array();
+    $rawLines = preg_split('/\R/u', $html);
 
-    // https://stackoverflow.com/a/8218649
-    if (!$dom->loadHTML(mb_convert_encoding($html, "HTML-ENTITIES", "UTF-8"))) {
-        throw new \UnexpectedValueException("The given HTML could not be parsed correctly.");
+    if ($rawLines === false) {
+        return false;
     }
 
-    $xpath = new \DOMXPath($dom);
-    $spans = $xpath->query("//span[contains(text(), '\n') or contains(text(), '\r\n')]");
+    foreach ($rawLines as &$rawLine) {
+        // If the previous line has been marked as "open", then we'll have something
+        // in our queue
+        if ($queuedPrefix !== '') {
+            $rawLine = $queuedPrefix . $rawLine;
+            $queuedPrefix = '';
+        }
 
-    /** @var \DOMElement $span */
-    foreach ($spans as $span) {
-        $closingTags = '';
-        $openingTags = '';
-        $curr = $span;
+        // Find how many opening `<span>` tags exist on this line
+        preg_match_all('/<span[^<>]+>/u', $rawLine, $regexWorkspace);
```

Comment:
In order to match `<span>`:

```suggestion
        preg_match_all('/<span[^>]*+>/u', $rawLine, $regexWorkspace);
```

This also uses a “possessive quantifier” to prevent backtracking which is unnecessary, I think.

And I think `[^>]` can be used instead of `[^<>]` to match the pattern below?

Reply:
Good catch, I for some reason thought `<` could end up in the attributes some how but that would need to be escaped anyways

Reply:
Actually, you're right that `<` can be in attributes, but so also can `>`. See https://jsbin.com/vitexad/edit?html,output

> ![image](https://user-images.githubusercontent.com/134745/102855594-c8289880-43d9-11eb-82cb-96a28fda201c.png)

This _works_ although and it is even valid HTML: 

> ![image](https://user-images.githubusercontent.com/134745/102855523-a6c7ac80-43d9-11eb-8bb0-27a8129893e5.png)

I think the question is whether or not `<` or `>` will ever show up in the `span` tags that highlight.php generates.

Reply:
I don't _think_ `<` and `>` would ever appear as class names since that's the only attribute highlight.php should ever generate; unless there's some really weird masochistic language grammar somewhere. I think it should be safe to ignore in this case

Reply:
Yeah, that makes sense.